blog/2018/12/12/dancer-and-minion/index.markdown

```diff
@@ -0,0 +1,299 @@
+was developed in Perl gave us the quickest time to implementation. 
+
+Scaling our existing setup was near impossible. It's not only not easy to get a handle on what resources are consumed by processes we've forked, but it was impossible to run the jobs on more than one server. 
+Starting over with Minion also gave us a much needed opportunity to clean up some code in sore need of refactoring. With a minimal amount of work, we were able to clean up our XML rendering code and make it work
+from Minion. This cleanup allowed us to more easily get information as to how much memory and CPU was consumed by an XML rendering job. This information is vital for us in planning future capacity.
+
+## Accessing Minion
+
+Since we are a Dancer shop, and not Mojolicious, a lot of convenience you'd get from Mojolicious for working with Minion isn't as available to us. Given we are also sharing some Minion-based
+code with our business models, we had to build some of our own plumbing around Minion:
+
+```
+package MyJob::JobQueue;
+
+use Moose;
+use Minion;
+
+use MyJob::Models::FooBar;
+with 'MyJob::Roles::ConfigReader';
+
+my @QUEUE_TYPES = qw( default InstantXML PayrollXML ChangeRequest );
+
+has 'runner' => (
+    is      => 'ro',
+    isa     => 'Minion',
+    lazy    => 1,
+    default => sub( $self ) {
+        $ENV{ MOJO_PUBSUB_EXPERIMENTAL } = 1;
+        Minion->new( mysql => MyJob::DBConnectionManager->new->get_connection_uri({ 
+            db_type => 'feeds', 
+            io_type => 'rw',
+        }));
+    },
+);
+
+sub has_invalid_queues( $self, @queues ) {
+    return 1 if $self->get_invalid_queues( @queues );
+    return 0;
+}
+
+sub get_invalid_queues( $self, @queues ) {
+    my %queue_map;
+    @queue_map{ @QUEUE_TYPES } = (); 
+    my @invalid_queues = grep !exists $queue_map{ $_ }, @queues;
+    return @invalid_queues;
+}
+
+sub run_job( $self, $args ) {
+    my $job_name = $args->{ name     } or die "run_job(): must define job name!";
+    my $guid     = $args->{ guid     } or die "run_job(): must have GUID to process!";
+    my $title    = $args->{ title    } // $job_name;
+    my $queue    = $args->{ queue    } // 'default';
+    my $job_args = $args->{ job_args };
+
+    die "run_job(): Invalid job queue '$queue' specified" if $self->has_invalid_queues( $queue );
+
+    my %notes = ( title => $title, guid  => $guid );
+
+    return $self->runner->enqueue( $job_name => $job_args => { notes => \%notes, queue => $queue });
+}
+
+1;
+
+```
+## Creating Jobs
+
+In our base model class (Moose-based), we would create an attribute for our job runner:
+```
+has 'job_runner' => (
+    is      => 'ro',
+    isa     => 'Empowered::JobQueue',
+    lazy    => 1,
+    default => sub( $self ) { return Empowered::JobQueue->new->runner; },
+);
+```
+And in the models themselves, creating a new queueable task was as easy as:
+```
+$self->runner->add_task( InstantXML => sub( $job, $request_path, $guid, $company_db, $force, $die_on_error = 0 ) {
```

Comment:
This line is a little long, maybe move to a newline for the sub at least? This is only stylistic of course.

Reply:
Done!